File: src/main/java/com/kishlaly/ta/analyze/tasks/blocks/one/Long_ScreenOne_MACD_LastAscending.java
package com.kishlaly.ta.analyze.tasks.blocks.one;

import com.kishlaly.ta.model.BlockResultJava;
import com.kishlaly.ta.model.SymbolDataJava;
import com.kishlaly.ta.model.indicators.IndicatorJava;
import com.kishlaly.ta.model.indicators.MACDJava;
import com.kishlaly.ta.utils.CollectionUtilsJava;
import com.kishlaly.ta.utils.LogJava;

import java.util.List;

import static com.kishlaly.ta.model.BlockResultCodeJava.HISTOGRAM_NOT_ASCENDING_SCREEN_1;
import static com.kishlaly.ta.model.BlockResultCodeJava.OK;

/**
 * the last histogram grows
 */
public class Long_ScreenOne_MACD_LastAscending implements ScreenOneBlockJava {
    @Override
    public BlockResultJava check(SymbolDataJava screen) {
        List<MACDJava> screen_1_MACD = (List<MACDJava>) screen.indicators.get(IndicatorJava.MACD);
        MACDJava screen_1_lastMACD = CollectionUtilsJava.getFromEnd(screen_1_MACD, 1);
        MACDJava screen_1_preLastMACD = CollectionUtilsJava.getFromEnd(screen_1_MACD, 2);

        boolean check2 = screen_1_lastMACD.getHistogram() > screen_1_preLastMACD.getHistogram();
        if (!check2) {
            LogJava.recordCode(HISTOGRAM_NOT_ASCENDING_SCREEN_1, screen);
            LogJava.addDebugLine("The histogram does not grow on the long-term screen");
            return new BlockResultJava(screen.getLastQuote(), HISTOGRAM_NOT_ASCENDING_SCREEN_1);
        }
        return new BlockResultJava(screen.getLastQuote(), OK);
    }
}

File: src/main/kotlin/com/kishlaly/ta/analyze/tasks/blocks/one/Long_ScreenOne_MACD_LastAscending.kt
package com.kishlaly.ta.analyze.tasks.blocks.one

import com.kishlaly.ta.model.BlockResult
import com.kishlaly.ta.model.BlockResultCode
import com.kishlaly.ta.model.SymbolData
import com.kishlaly.ta.model.indicators.Indicator
import com.kishlaly.ta.model.indicators.MACD
import com.kishlaly.ta.utils.CollectionUtils
import com.kishlaly.ta.utils.Log

/**
 * the last histogram grows
 */
class Long_ScreenOne_MACD_LastAscending : ScreenOneBlock {
    override fun check(screen: SymbolData): BlockResult {
        val screen_1_MACD = screen.indicator(Indicator.MACD) as List<MACD>
        val screen_1_lastMACD = screen_1_MACD.last()
        val screen_1_preLastMACD = CollectionUtils.getFromEnd(screen_1_MACD, 2)

        val check = screen_1_lastMACD.histogram > screen_1_preLastMACD.histogram
        if (!check) {
            Log.recordCode(BlockResultCode.HISTOGRAM_NOT_ASCENDING_SCREEN_1, screen)
            Log.addDebugLine("The histogram does not grow on the long-term screen")
            return BlockResult(screen.lastQuote, BlockResultCode.HISTOGRAM_NOT_ASCENDING_SCREEN_1)
        }
        return BlockResult(screen.lastQuote, BlockResultCode.OK)
    }
}